node/src/actors/chain_manager/mod.rs

```diff
@@ -1104,6 +1313,65 @@ impl ChainManager {
         let epoch = self.current_epoch.unwrap();
         self.sync_waiting_for_add_blocks_since = Some(epoch);
     }
+
+    fn process_blocks_batch(
+        &mut self,
+        ctx: &mut Context<Self>,
+        sync_target: &SyncTarget,
+        blocks: &[Block],
+    ) -> (bool, usize) {
+        let mut batch_succeeded = true;
+        let mut num_processed_blocks = 0;
+
+        for block in blocks.iter() {
+            num_processed_blocks += 1;
+
+            if let Err(e) = self.process_requested_block(ctx, block.clone()) {
+                log::error!("Error processing block: {}", e);
+                self.initialize_from_storage(ctx);
+                log::info!("Restored chain state from storage");
+                batch_succeeded = false;
+                break;
+            }
+
+            let beacon = self.get_chain_beacon();
+            show_sync_progress(beacon, &sync_target, self.epoch_constants.unwrap());
+        }
+
+        (batch_succeeded, num_processed_blocks)
+    }
+
+    fn process_first_batch(
+        &mut self,
+        ctx: &mut Context<ChainManager>,
+        sync_target: &SyncTarget,
+        blocks: &[Block],
+    ) -> (bool, usize) {
+        let (batch_succeeded, num_processed_blocks) =
+            self.process_blocks_batch(ctx, &sync_target, &blocks);
+
+        if !batch_succeeded {
+            log::error!("Received invalid blocks batch");
+            self.sm_state = StateMachine::WaitingConsensus;
+            self.sync_waiting_for_add_blocks_since = None;
+        }
+
+        (batch_succeeded, num_processed_blocks)
+    }
+
+    fn superblock_consolidation_is_needed(
+        &self,
+        sync_target: &SyncTarget,
+        superblock_period: u32,
+    ) -> Option<u32> {
+        if sync_target.superblock.checkpoint
+            == self.chain_state.superblock_state.get_beacon().checkpoint
```

Comment:
```suggestion
            != self.chain_state.superblock_state.get_beacon().checkpoint
```

Reply:
I changed

Reply:
I changed the `if` and `else` branches :)